File: 17 09 25/mergeintervals.ex
defmodule MergeIntervals do
  def merge(intervals) do
    intervals
    |> Enum.sort_by(&elem(&1, 0))
    |> Enum.reduce([], fn {s, e}, acc ->
      case acc do
        [{ls, le} | rest] when s <= le -> [{ls, max(le, e)} | rest]
        _ -> [{s, e} | acc]
      end
    end)
    |> Enum.reverse()
  end

  # ---- Example Uses ----
  def demo do
    IO.inspect merge([{1, 3}, {2, 6}, {8, 10}, {15, 18}])
    IO.inspect merge([{1, 4}, {4, 5}])
    IO.inspect merge([{1, 2}, {3, 4}, {5, 6}])
    IO.inspect merge([{1, 10}, {2, 3}, {4, 5}, {6, 7}])
  end
end

File: 17 09 25/mergeintervals.erl
-module(merge_intervals).
-export([merge/1, demo/0]).

merge(Intervals) ->
    Sorted = lists:sort(fun({A,_},{B,_}) -> A < B end, Intervals),
    merge_sorted(Sorted, []).

merge_sorted([], Acc) -> lists:reverse(Acc);
merge_sorted([{S,E}|Rest], []) -> merge_sorted(Rest, [{S,E}]);
merge_sorted([{S,E}|Rest], [{LS,LE}|AccTail]) ->
    case S =< LE of
        true  -> merge_sorted(Rest, [{LS, max(LE,E)}|AccTail]);
        false -> merge_sorted(Rest, [{S,E},{LS,LE}|AccTail])
    end.

